specs/packagemanager/PackageManagement.md

```diff
@@ -0,0 +1,1136 @@
+- [6. API Details](#6-api-details)
+
+# 2. Background
+
+Windows sports the ability to deploy and manage software as MSIX packages via WinRT APIs in the
+Windows.Management.Deployment namespace. These APIs provide the means to install, update, uninstall,
+enumerate and otherwise manage packages.
+
+The package management APIs were originally introduced in Windows 8 and have continuously expanded
+to meet MSIX's growing demands over the past decade. However, as one of the earliest WinRT APIs they
+include some patterns out of step with current API practices and recommendations. Windows App SDK
+offers the opportunity to provide a new generation of package management APIs in line with the
+latest patterns and recommendations for an improved developer experience as well as functional
+enhancements and improved runtime efficiencies.
+
+Microsoft-internal task [45952398](https://task.ms/45952398)
+
+# 3. Description
+
+This API provides enhanced access to Windows' package management capabilities, focusing on the
+following scenarios:
+
+* Stage a package
+* Register a package
+* Add/Update a package
+* Remove a package
+* Repair a package
+* Reset a package
+* Provision a package
+* Deprovision a package
+
+Additional functionality includes:
+
+* IsReady -- Is a package ready for use?
+* EnsureIsReady -- Is a package ready for use and, if not, make it so
+* IsPackageRegistrationPending -- Is there an update waiting to install?
+* PackageSets -- Batch operations
+* PackageRuntimeManager -- Batch operations for use at runtime via Dynamic Dependencies
+* Usability -- Quality-of-Life enhancements
+
+## 3.1. API Structure
+
+Methods to drive deployment activity typically follow the pattern:
+
+    <verb>Package[Set][By<Type>][Suffix]Async(target, options)
+
+The following verbs are supported:
+
+* Is...Ready
+* Ensure...IsReady
+* Add
+* Stage
+* Register
+* Remove
+* Repair
+* Reset
+* Provision
+* Deprovision
+
+These methods accept their target package(s) as their first parameter. This parameter can be an
+various types of information including:
+
+* Filename
+* Path
+* PackageFamilyName
+* PackageFullName
+* Uri
+
+**NOTE:** Methods with a target of a URI are named ...ByUri....
+
+**NOTE:** Methods with a target of a PackageFamilyName are named ...ByPackageFamilyName....
+
+**NOTE:** Methods with a target of a PackageFamilyName are named ...ByPackageFullName....
+
+**NOTE:** Methods with no ...By<Type>... qualifier accept a string which can contain one or more types
+of identifiers. These can vary for different verbs. See the per-method documentation for the
+specific target types supported by each method.
+
+These methods accept options as a matching `<verb>Package[Set]Options` type, e.g.
+`AddPackageAsync(string packageUri, AddPackageOptions options)`.
+
+The following table shows the supported permutations of verbs and targets:
+
+|Verb         | Path | Filename | PackageFamilyName | PackageFullName | file:  | http(s): | ms-uup: | PackageSet |
```

Comment:
What's the difference between "path", "filename" and "file:"?
In OS PackageManager APIs these are interchangeable.

Reply:
file: is a URI 
path is a directory
file is a filename

**Examples**

* URI vs file = file://c:/foo/bar.msix vs c:\foo\bar.msix
* path vs file = c:\blah vs c:\blah\appxmanifest.xml

OS supports `Register(file://c:/blah/appxmanifest.xml)`. WAS will also support `Register(file://c:/blah)`.

Add and Stage have some similar permutations e.g. `Add(c:\blah, options.StageInPlace=true)`.

Table was wrong, Add+Stage support Path today. Fixed. Checking on Register